silq/instrument_interfaces/rohde_schwarz/SGS100A_interface.py

```diff
@@ -0,0 +1,411 @@
+        assert pulse.power is not None, "Pulse must have power defined"
+        return super().target_pulse(pulse, interface, **kwargs)
+
+    def get_additional_pulses(
+        self,
+        interface: InstrumentInterface,
+        frequency: float,
+        IQ_modulation: bool,
+        power: float,
+    ):
+        if not IQ_modulation:
+            return []
+        else:  # interface.IQ_modulation() == 'on'
+            attenuation = self.pulse.power - power
+            amplitude = 10.0 ** (attenuation / 20)
+            assert amplitude <= 1.01, f"IQ amplitude larger than 1: {amplitude}"
+
+            frequency_IQ = self.pulse.frequency - frequency
+            additional_pulses = [
+                SinePulse(
+                    name="sideband_I",
+                    t_start=self.pulse.t_start - interface.envelope_padding(),
+                    t_stop=self.pulse.t_stop + interface.envelope_padding(),
+                    frequency=frequency_IQ,
+                    amplitude=amplitude,
+                    phase=self.pulse.phase,
+                    connection_requirements={
+                        "input_instrument": interface.instrument_name(),
+                        "input_channel": "I",
+                    },
+                ),
+                SinePulse(
+                    name="sideband_Q",
+                    t_start=self.pulse.t_start - interface.envelope_padding(),
+                    t_stop=self.pulse.t_stop + interface.envelope_padding(),
+                    frequency=frequency_IQ,
+                    phase=self.pulse.phase - 90,
+                    amplitude=amplitude,
+                    connection_requirements={
+                        "input_instrument": interface.instrument_name(),
+                        "input_channel": "Q",
+                    },
+                ),
+            ]
+            return additional_pulses
+
+
+class FrequencyRampPulseImplementation(PulseImplementation):
+    pulse_class = FrequencyRampPulse
+
+    def target_pulse(self, pulse, interface, **kwargs):
+        assert pulse.power is not None, "Pulse must have power defined"
+        return super().target_pulse(pulse, interface, **kwargs)
+
+    def get_additional_pulses(
+        self,
+        interface: InstrumentInterface,
+        frequency: float,
+        IQ_modulation: bool,
+        power: float,
+    ):
+        assert IQ_modulation
+
+        attenuation = self.pulse.power - power
+        amplitude = 10.0 ** (attenuation / 20)
+        assert amplitude <= 1.01, f"IQ amplitude larger than 1: {amplitude}"
+
+        additional_pulses = [
+            FrequencyRampPulse(
+                name="sideband_I",
+                t_start=self.pulse.t_start,
+                t_stop=self.pulse.t_stop,
+                frequency_start=self.pulse.frequency_start - frequency,
+                frequency_stop=self.pulse.frequency_stop - frequency,
+                amplitude=amplitude,
+                phase=0,
```

Comment:
a bit unfair to neglect the phase of the pulse, but it is depends on the application

Reply:
fixed